joomla/joomla-cms
# [5.0][RFC] Joomla Dialog (popup) script. Free, without CC and registration

### Summary of Changes

This is implementation of own popup/modal/dialog independend from bootstrap and other traps.
As continue of https://github.com/joomla/joomla-cms/discussions/32473 discussion.

At current point of time the PR provides only popup implementation, the script should be ready to use.
And the integration will come.

~~I made it for 4.x to be able easily test on "production" codebase. However main target (If I ever finish it) probably > 5.x.~~

The PR does not provide a separate stylesheet for popup, the styling should be done within template (example can find in _modal.scss of this PR).
(At bare it use native dialog styling)



### What and why is Joomla Dialog

The script allow to create a popup/dialog programaticaly, that in oposite to current Bootstrap implementation, when modal should be prerendered in HTML. This allows more simple integration on client side without involving server side.

It basicaly a wrapper around native `dialog` element.

Joomla Dialog script allows to display dialog with following content: 
- `inline` - text or html content;
- `iframe` - embed/remote content, in iframe;
- `ajax` - same as `inline` but the content loaded through ajax request;
- `image` - image lightbox;

Also it provides helper methods for `alert()` and `confirm()` dialogs.
Provides helper to bind page button/anchor to show the popup, for basic stuff, without extra js needed.

### Usage examples
```php
<?php $wa->useScript('joomla.dialog'); ?>
```
```javascript
// Inline 
const dialog = new JoomlaDialog({
  textHeader: 'The header',
  popupContent: '<p class="p-3">Popup content text</p>',
});
dialog.show();

// IFrame
const dialog = new JoomlaDialog({
  popupType: 'iframe',
  textHeader: 'The header',
  src: 'index.php?option=com_content&view=articles&tmpl=component&layout=modal',
});
dialog.show();

// Ajax
const dialog = new JoomlaDialog({
  popupType: 'ajax',
  textHeader: 'The header',
  src: 'index.php?option=com_content&view=articles&tmpl=component&layout=modal',
});
dialog.show();

// Image
const dialog = new JoomlaDialog({
  popupType: 'image',
  src: '../images/headers/walden-pond.jpg',
});
dialog.show();

```

Any property can be set as instance property or in class constructor. This two is equal:

```javascript
// Proprty in class constructor
const dialog = new JoomlaDialog({
  textHeader: 'The header',
  popupContent: '<p class="p-3">Popup content text</p>',
});

// Proprty in to class instance
const dialog = new JoomlaDialog();
dialog.textHeader = 'The header';
dialog.popupContent = '<p class="p-3">Popup content text</p>';
```

Other options:
```javascript
// Optional sizing:
dialog.width = '80vw';
dialog.height = '80vh';

// Definig your own buttons:
dialog.popupButtons = [
  { label: 'Yes', onClick: () => dialog.close() },
  { label: 'No', onClick: () => dialog.close(), className: 'btn btn-outline-danger ms-2' },
];
```

#### Alert and Confirm usage:
```javascript
// Alert
JoomlaDialog.alert('Chase ball of string eat plants, meow')
  .then(() => { 
    console.log('All done'); 
  });

// Confirmation
JoomlaDialog.confirm('Cheese on toast airedale the big cheese?')
  .then((result) => { 
    console.log(result ? 'Okay' : 'Not okay'); 
  });
```

#### Binding button/anchor for basic stuff

```php
<?php $wa->useScript('joomla.dialog-autocreate'); ?>

<button class="btn btn-primary" type="button"
    data-joomla-dialog='{"popupType": "iframe", "width":"80vw", "height": "80vh", "src":"index.php?option=com_content&view=articles&tmpl=component&layout=modal"}'>Click</button>

<button class="btn btn-primary" type="button"
    data-joomla-dialog='{"popupType": "inline", "src":"#popupText", "width": "fit-content", "height": "fit-content"}'>Click</button>

<a href="index.php?option=com_content&view=articles&tmpl=component&layout=modal"
  data-joomla-dialog class="btn btn-outline-primary">Click</a>
<a href="#popupText" data-joomla-dialog class="btn btn-outline-primary">Click</a>

<template id="popupText"><p class="p-3">Popup content text</p></template>
```

### Testing Instructions

Can test any of examples above.
Apply patch, run `npm install`.
Then add `$wa->useScript('joomla.dialog');` in to index.php of the admin template.
And one of example js/html wherever you want. Eg. in to index.php (will show 2 popup on load):

```html
<script type="module">
import JoomlaDialog from 'joomla.dialog';

JoomlaDialog.alert('Chase ball of string eat plants, meow')
  .then(() => { 
    console.log('All done'); 
  });

JoomlaDialog.confirm('Cheese on toast airedale the big cheese?')
  .then((result) => { 
    console.log(result ? 'Okay' : 'Not okay'); 
  });
</script>
```



### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: TBD
- [x] Pull Request link for manual.joomla.org: https://github.com/joomla/Manual/pull/99

### Issues for reference

- #32473
- #39344
- #39964
- #39449 
- #39609
- #39416
- #40202

@dgrammatiko I have stole some of your ideas from #39344, I hope you will not sue me :)
